Design Document Bucket Path (#105)
exasol/bucketfs-python: doc/design/bucketpath.rst

==============================
Design Document Bucket Path(s)
==============================

Introduction
============

- Purpose of the document / Problem description
- Definitions and Acronyms

Overview
========

Design Goals
============

Path Design
===========

Future Considerations
=====================

Sudo Code Outline
=================

.. code-block:: python

    from exasol.bucketfs import PathBuilder

>     Path = PathBuilder(credentials)

Reviewer: is this only for serialization or is this the factory?

Author: This is the factory, it is supposed to bring together the serialization `as_uri()` and the credentials (and/or other outside information)

>     udf_path = Path("bfsl://some/local/path/file.tar.gz")

Reviewer: for what stands bfsl

Author: Yes for the lack of a better name the l stands for local. The major question here would be is a local path different from a local path in the UDF?

>     http_bucket_path = Path("bfs://127.0.0.1:8888/service/bucket/some/file.tar.gz")

Reviewer: if I understand that correct, the might look then like that for a concrete example

http_bucket_path = Path("bfs://127.0.0.1:8888/bfsdefault/default/some/file.tar.gz")

>     http_bucket_path = Path("bfs://127.0.0.1:8888/service/bucket/some/file.tar.gz")

Reviewer: We will need also later a SaaS url for that

>     https_bucket_path = Path("bfss://127.0.0.1:8888/service/bucket/some/file.tar.gz")

Reviewer: I am actually not sure if we should point to files at all, because Bucket and what ever we define for DirectoryBucket can also only point to a Directory. Except, we want to go futher and actually point to a File Abstraction.

Author: We don't need to but it should be possible, I think `Path` is the right abstraction.
Given that we can discuss on it's feature set.

Reviewer: Yeah, after reading the Path class, it probably makes sense to point to files. If I think about the usage, we actually might want to point to a file, such that a UDF uses a specific file for example.

>     chroot_path = Path("bfss://127.0.0.1:8888/service/bucket/some/sub/subsub/file.tar.gz?chroot=/some/sub/")

Reviewer: not yet sure we need the chroot, a path to a directory would be enough I think.

Author: I am all on board with that, less is more for starters. But I think in the past we where discussing some use cases to make sure one can't traverse above a "pinned" directory.

Reviewer: Yeah, we have this case for example with temporary directories, where the user should not write to the parent, because then it doesn't get cleaned. At the moment, this is done by not having a function to go to the parent of a location and probably by convention. So, we probably need it, when we allow the normal path to go to its parent.

.. code-block:: python

    from typing import Protocol

    # interface / protocol or abc
    # Note: should mimic pathlib.Path so it can easily understood and adopted. 
    class Pathlike(Protocol):

        @property
        def name:
            ...

        @property
        def suffix:
            ...

        @property
        def root:
            ...

        def as_uri():
            ...

        def exists():
            ...

        def is_dir():
            ...

        def is_file():
            ...

        def read_text():
            ...

        def read_bytes():
            ...

        def write_text():
            ...

>         def write_bytes():

Reviewer: this we need to orient more on what we are doing in the bucket, aka download and upload, because it is not simple reads and writes and especially we need to be able to read and write in chunks, but we can't support partial changes.

Reviewer: Agree. I thought about the open two, but I am not sure if we can implement a FileObj, but in the end it doesn't need to be a full and real FileObj. Maybe, we don't allow seek and things and only implement the read and write.

Reviewer: I now also remember what the issue with the FileObj was. We need to use a generator or iterator in requests.put to feed the data in chucks. There we have a pull principle for the data. However, FileObj write implements a push principle. That means, you probably need to spawn thread where the requests.put needs to run in and connect it with a queue to the write.

Author: I agree`open` does like impose additional considerations, so I would put that up for ... "potential improvements later".

Reviewer: so, I would suggest that we add additional to read_* and write_* the function we use in the bucket, they are already known. Later, we can think about open.

            ...

        def unlink():
            ...

        def rmdir():

>             ...

Author: In Bucketfs they don't but this abstraction, shall provide the means of "Directories" right. 
If we want to be able to delete all bellow that dir I think we should add it, if we don't want to support directory deletion that's fine too. 

Regarding the unlink I am not sure it would help understand `pathlib.Path`  users more easily whats going on, on the other hand it may isn't obvious for non FS people so an `rm` or `delete` may would be better.

>             ...

Reviewer: Yeah, I think, we need to iterate over this and check the exact meaning of these functions.

>             ...

Reviewer: I had a look at unlink and rmdir. Unlink is ok, because it only works on files. However, rmdir only works for empty directories which doesn't exist in bucketfs, because as soon a directory gets empty in the bucketfs it automatically vanish.

>             ...

Reviewer: ok to stay compatible with the API, we can implement rmdir as you suggested. And, yes we probably need something like rmtree

        def join()
            ...

        def walk():
            ...

        def iterdir():
            ...

        # overload / for joining
        def __truediv__():
            ...

    # Note: needs to support Pathlike
    class BucketPath:
        
        protocol = ['bfs', 'bfss']

        def __init__(bucket: Bucket, path):
            ...

>     # QUESTION: UdfPath? Do we need to distuinguish between localpaths and localpaths within udf?

Reviewer: UDFPath can't write, besides that both operate on local file system. Only the local path we create might be different. For testing it points to a temp dir and for udfs in points to `/buckets/<service>/<bucket>/...`

Author: TL;DR: Local paths are working exactly the same except they having a different root? (So the Chroot may still come in handy here)

Reviewer: Actually a good point. We could setup the same structure as in the UDF in a temporary directory. For example:

`/tmp/tmpdir/buckets/default/path/in/bucket`

==============================
Design Document Bucket Path(s)
==============================

Problem Description
===================

> Users of the bucketfs file system need to use it in various diffrent contexts like, from the outside of the
> DB interacting with bucketfs, from within the DB when accssing bucketfs path from within udfs.
> Also common actions/tasks like listing a directory are pretty tidous when just interacting with
> the bucket api due to the fact that it does not know the concept of direcotries. So in 
> order to simplify and streamline frequently used path operations and also provide a uniform 
> interface accross the actual system (local path, http, ...) behind the bucketfs path we need
> to have an abstraction for the user.
> 
> The BucketFS filesystem is an integral part of sharing and accessing data for its users.
> However, the current user experience with BucketFS presents several challenges, particularly in terms of versatility and ease of use across different contexts. Users interact with BucketFS both externally, from outside the database, and internally, within database operations such as accessing paths from within User-Defined Functions (UDFs). This dual mode of interaction introduces complexity and inefficiency, especially for common filesystem operations.

Reviewer 2: ```suggestion
Users of the BucketFS file system need to use it in various diffrent contexts like, from the outside of the
DB interacting with bucketfs, from within the DB when accessing BucketFS paths from within UDFs.
Also common actions/tasks like listing a directory are pretty tedious when just interacting with
the BucketFS API due to the fact that it does not know the concept of directories. So in 
order to simplify and streamline frequently used path operations and also provide a uniform 
interface accross the actual system (local path, http, ...) behind the BucketFS path we need
to have an abstraction for the user.

The BucketFS filesystem is an integral part of sharing and accessing data for its users.
However, the current user experience with BucketFS presents several challenges, particularly in terms of versatility and ease of use across different contexts. Users interact with BucketFS both externally, from outside the database, and internally, within database operations such as accessing paths from within User-Defined Functions (UDFs). This dual mode of interaction introduces complexity and inefficiency, especially for common filesystem operations.
```

Reviewer 2: I don't understand the following sentence: 

> The BucketFS filesystem is an integral part of sharing and accessing data for its users.

Reviewer 2: Additionally, I have the feeling that both paragraphs of the problem description are more or less telling the same in a somewhat repetitive manner. Could we maybe shorten and simplify the introduction?

Challenges with Current BucketFS Interactions
+++++++++++++++++++++++++++++++++++++++++++++

1. **Contextual Versatility:**

>    Users face difficulties when switching between different operational contexts, such as external access (http) versus internal access (local pahts, during UDF execution). The lack of a seamless transition between contexts hinders productivity and introduces additonal code complexities.

Reviewer 2: ```suggestion
   Users face difficulties when switching between different operational contexts, such as external access (HTTP) versus internal access (local paths, during UDF execution). The lack of a seamless transition between contexts hinders productivity and introduces additonal code complexities.
```

2. **Directory Operations:**

>    BucketFS inherently lacks the concept of directories as understood in traditional file systems. This absence complicates tasks like listing directory contents, making what should be simple actions cumbersome and time-consuming. Users are forced to interact with a lower-level bucket API for directory-like operations, which is not only tedious but also counterintuitive.

Reviewer 2: ```suggestion
   BucketFS inherently lacks the concept of directories as understood in traditional file systems. This absence complicates tasks like listing directory contents, making what should be simple actions cumbersome and time-consuming. Users are forced to interact with a lower-level BucketFS API for directory-like operations, which is not only tedious but also counterintuitive.
```

- **Focus on Essential Functionalities:** Although inspired by `pathlib`, the proposed abstraction will be streamlined to include only the functionalities necessary for effective BucketFS management. It will diverge from `pathlib` in areas specific to BucketFS's architecture and use cases, ensuring a tailored approach.

- **Enhance Filesystem Operations:** The abstraction is designed to facilitate common filesystem tasks such as directory listings and file operations, addressing the current lack of a unified method for such interactions in BucketFS. This enhancement aims to standardize the way users interact with the filesystem, regardless of the backend system.

The development of a path abstraction layer, while inspired by `pathlib`, is specifically designed to meet the unique interaction requirements of BucketFS. This proposed solution focuses on the practical needs of BucketFS users, aiming to streamline their workflow and improve efficiency in managing filesystem tasks.

Desgin
======

Design Goals
++++++++++++

> The primary goal of this design is to create an abstraction that simplifies working with the bucketfs file system and it's usages.

Reviewer 2: ```suggestion
The primary goal of this design is to create an abstraction that simplifies working with the BucketFS file system and its usages.
```

Additionaly we wan't to maintain compatibility with the intuitive and widely used `pathlib` interface(s) where possible.

This abstraction should:

> - Simplify interacting with bucketfs paths, by providing implentations for common tasks.

Reviewer 2: ```suggestion
- Simplify interacting with BucketFS paths, by providing implentations for common tasks.
```

- Provide a way to persist and/or share path information accross processes and systems.
- Reduce the learning curve for users familiar with the `pathlib` interface.

> - Make sure the behaviour follow `pathlib` wherever possible.

Reviewer 2: ```suggestion
- Make sure the behaviour follows `pathlib` wherever possible.
```

- For significant semantic deviations, opt for unique, clear names that avoid confusion with ``pathlib`` terminology.

.. Attention::

   The subsequent subsections include code snippets intended primarily for the implementer's reference.
   Therefore, comments and docstrings within the sudo code may need adaptation for the actual implementation.

Pathlike
--------

.. code-block:: python

    from typing import Protocol

    class Pathlike(Protocol):

        @property
        def name:
            """
            A string representing the final path component, excluding the drive and root, if any.
            """

        @property
        def suffix:
            """
            The file extension of the final component, if any.
            """

        @property
        def root:
            """
            A string representing the root, if any.
            """

        @property
        def parent:
            """
            The logical parent of this path.
            """

        def as_uri():
            """
            Represent the path as a file URI. Can be used to reconstruct the location/path.
            """

>         def exists():

Author: Paths are abstract, similar to file-based paths where a non-existent path can be created using write or touch. For pathlike objects, write can bring a path into existence.

            """
            Return True if the path points to an existing file or directory.
            """

>         def is_dir():

Author: A `Path` can even though BucketFS does not have the possibility to explicitly creating a specific empty directory.

            """
            Return True if the path points to a directory, False if it points to another kind of file.
            """

        def is_file():
            """
            Return True if the path points to a regular file, False if it points to another kind of file.
            """

        def read(chunk_size: int = 8192) -> Iterable[ByteString]:
            """
            TBD
            """

        def write(data: ByteString | BinaryIO | Iterable[ByteString]):
            """
            TBD
            """

>         def rm():

Reviewer 2: Please note that currently there is no way to detect when a delete operation is finished, i.e. propagated to all notes of a database cluster.

Author: Thanks for reminding me on that, still a means of deleting files will be required.

            """
            If exists() and is_file yields true for this path, the path will be deleted,
            otherwise exception will be thrown.
            """

        def rmdir(recursive: bool = False):
            """
            TBD
            """

        def joinpath(*pathsegements)
            """
            Calling this method is equivalent to combining the path with each of the given pathsegments in turn
            """

        def walk():
            """
            TBD
            """

        def iterdir():
            """
            TBD
            """

        # Overload / for joining, see also joinpath or `pathlib.Path`.
        def __truediv__():
            """
            TBD
            """

Concrete Paths (Backends)
-------------------------

Implement the ``Pathlike`` protocol for each specific backend. Concrete paths must include:

- A ``protocol`` member indicating the associated protocol.
- Backend-specific methods for creation.
- Implementations for all methods and properties required by the ``Pathlike`` protocol

> Each backend must implement the ``as_uri`` method in a way that the location is clearly identifiable.

Reviewer 2: I am not sure whether this can be provided in all cases. 
It might depend on the definition of "uri".
What is an URI in this context?

Reviewer 2: If a URI is understood as a string identifying a bucket or path in a bucket from outside, then some parts of the URI might be unknown to a UDF.

Author: For details on this please see the example at the bottom (factory section).

.. code-block:: python

    # Attention: needs to provide/implment Pathlike protocol
    class BucketPath:
        """

>         Provides access to a bucket path servered via http or https.

Reviewer 2: ```suggestion
        Provides access to a bucket path served via http or https.
```

        """
        
        # uri protocol specifies associated with this class
        protocol = ['bfs', 'bfss']

        def __init__(bucket: Bucket, path: str):
        """
        Creates a new BucketPath.

        Args:
            bucket: used for accssing and interacting with to the underlying bucket.
            path: of the file within the bucket.
        """
            
        # Pathlike functionalities
        # ...

    # Attention: needs to provide/implment Pathlike protocol
    class LocalPath:
        """

>         Provides access to a bucket path servered local file system paths.

Reviewer 2: Does "local" mean "inside the UDF"?

>         Provides access to a bucket path servered local file system paths.

Reviewer 2: ```suggestion
        Provides access to a bucket path served local file system paths.
```

                *args: TBD
                *kwargs: TBD

            Returns:
                A `PathBuilder` object which can be aliased as `Path` for creating paths based on uris.
            """
            pass

        def __call__(uri: str) -> Pathlike:
            """
            Creates a new Path (BucketPath, LocalPath, ...) based on the provided uri.

            Args:
                uri: for which a Path (Pathlike) object should be created.

            Returns:
                A Pathlike object for the given uri.
            """
            # type: LocalPath, BucketPath, Chroot ...
            # 
            # Note: based on the uri the factory should assemble the apropriate Pathlike object.
            # E.g.:
            type = _determine_type(path)
            facories = {
                "udf" = _create_udf_path,
                "bfs" = _create_bucket_path,
                "chroot" = _create_chroot_path,
            }
            factory = factory[type]
            return factory(args)

Examples
^^^^^^^^

.. code-block:: python

    from exasol.bucketfs import PathBuilder

    Path = PathBuilder(credentials)

>     # Create differnet kind's of bucketfs paths

Reviewer 2: ```suggestion
    # Create different kinds of bucketfs paths
```

            """

        def read(chunk_size: int = 8192) -> Iterable[ByteString]:
            """
            Read the content of a the file behind this path.

            Only works for pathslike objects which return True for `is_file()`.

            Args:
                chunk_size: which will be yieled by the iterator.

            Returns:
                Returns an iterator which can be used to read the contents of the path in chunks.

            Raises:
                NotAFileError: if the pathlike object does not point to a file.
                FileNotFoundError: If the file does not exist.
            """

        def write(data: ByteString | BinaryIO | Iterable[ByteString]):
            """
            Writes data to a this path. 

            After successfully writing to this path `exists` will yield true for this path.

            Args:
                data: which shall be writen to the path.

            Raises:
                FileExistsError: if the file already exists.
                NotAFileError: if the pathlike object is not a file path.

                Note: 

>                 If you still need to write the new content, please delete the current contents first,

Reviewer: this I don't get, overwriting is possible

Author: This is a challenging design decision. Implementing overwriting functionality is feasible; however, it necessitates the deletion of the existing file in the Bucket. This process complicates the provision of a stable API, as file deletion requests do not immediately remove the file from BucketFS.

Reviewer: overwrite should work without problems

                for futher details see `rm()`.
            """

        def rm():
            """
            Remove this file.

            Note:
                If `exists()` and is_file yields true for this path, the path will be deleted,
                otherwise exception will be thrown.

            Raises:
                FileNotFoundError: If the file does not exist.
            """

        def rmdir(recursive: bool = False):
            """
            Removes this directory.

            Note: In order to stay close to the pathlib based `rmdir` the default won't delete

>                   non empty direcotries, in order to to so recus

Reviewer: recus?, seems something is missing

Reviewer: ```suggestion
                  non empty direcotries, in order to do so recus
```

            Args:
                recursive: if true the entrie directory and it's entire contents (files and subdirs)

>                            will be delted. If false and the directory is not empty an error will be thrown.

Reviewer: ```suggestion
                           will be deleted. If false and the directory is not empty an error will be thrown.
```

            Returns:
                A new pathlike object pointing the combined path.
            """

        def walk() -> Tuple[Pathlike, List[str], List[str]]:
            """
            Generate the file names in a directory tree by walking the tree either top-down or bottom-up.

            Note:
                Try to mimik https://docs.python.org/3/library/pathlib.html#pathlib.Path.walk as closely as possible,
                except the functionality associated with the parameters of the `pathlib` walk.

            Yields:
                A 3-tuple of (dirpath, dirnames, filenames).
            """

        def iterdir() -> Generator[Pathlike, None, None]:
            """
            When the path points to a directory, yield path objects of the directory contents.

            Note:
                For file like paths it will yield nothing.

            Yields:
                All direct childs of the pathlike object.
            """

        # Overload / for joining, see also joinpath or `pathlib.Path`.
        def __truediv__():
            """
            """

> 

Reviewer: w could think about glob and rglob, these functions are really useful for lazy people like me

Author: For now I would focus on the core functionality for now, if it turns out that this is frequently used/needed we can add it at a later point in time.

Concrete Paths (Backends)
-------------------------

Implement the ``Pathlike`` protocol for each specific backend. Concrete paths must include:

> - A ``protocol`` member indicating the associated protocol.

Reviewer: this doesn't mean a python protocol, you mean a connection protocol, I guess

Author: yes, basically the protocol part of an uri/url.

- Backend-specific methods for creation.
- Implementations for all methods and properties required by the ``Pathlike`` protocol

Each backend must implement the ``as_uri`` method in a way that the location is clearly identifiable.

.. code-block:: python

    # Attention: needs to provide/implment Pathlike protocol
    class BucketPath:
        """
        Provides access to a bucket path served via http or https.
        """
        
        # uri protocol specifies associated with this class
        protocol = ['bfs', 'bfss']

        def __init__(bucket: Bucket, path: str):
        """
        Creates a new BucketPath.

        Args:
            bucket: used for accssing and interacting with to the underlying bucket.
            path: of the file within the bucket.
        """

>             

Reviewer: we might also need a function in PathLike that returns us the path in the UDFs, because often we let the user provide us the information for external access + the bucketfs service name + credentials and use this for first uploading stuff and then get the UDF path to read the uploaded stuff from local file system in the UDF

Author: From my pov such a function should not be in `Pathlike`, this would break the abstraction by exposing backend specifics. I would recommend to start out with providing a simple utility function which can do such a thing on the side.

Author: All necessary information is contained within the URI, accessible via as_uri(). Therefore, it's feasible to introduce a standalone function, as_local, to handle this. Integrating it into the "interface" later will be straightforward and non-disruptive. Altering the interface signature, in contrast, is of significantly greater impact.

Reviewer: extra function which converts http uri to local uri sounds good

            chroot: the path like object shall be rectricted/pinned to.

        Returns:
            Creates a new Chroot protected path like object.
        """
        # Pathlike functionalities
        # Note: functions like ``parent`` should stop at the new root
        # ...

    # Attention: needs to provide/implment Pathlike protocol
    class ReadOnly:
        """
        Modifies a pathlike object so it will be readonly.
        """

        def __init__(path: Pathlike):
            """
            Create a new ReadOnly Pathlike object.

            Args:
                path: like object which will be write proected (readonly).

            Returns:
                A path like object whith write proection.
            """
        
        # Pathlike functionalities
        # Note: Non readonly actions should throw an exception
        # ...

Factory & Builders
------------------

.. code-block:: python

    def PathBuilder:

        def __init__(credentails_store, *args, **kwargs):
            """
            Initalizes the factory with settings which are required besides the actual path uri.

            Args:

>                 credentail_store: for accessing buckets, see Service.

Reviewer: the credential_store probably will get later, a bit criminal, because you might remember, I that,that we have a use case, where I need to wipe all the credential information from the Bucket Objects/Pathlike Objects before serializing them and later reinject the credentials in the deserialized Bucket Objects/Pathlike without that the part that serializes and deserialized has direct references to the Bucket Objects/Pathlike 

For example:
```
def step1(credentials)
    credential_store = CredentialStore(credentials)
    Path = PathFactory(credential_store, ...)
    component = Component(Path)
    component.do_something() -> this creates PathLikes internallly which get save in the component
    credential_store.clean()
    serialize(component, credential_store)

def step2(credentials)
    component = deserialize()
    component, credential_store = CredentialStore(...)
    credential_store.repopulate(credentials)
    component.do_something_else() -> uses the stored PathLikes
  

```

Author: From my pov the serialization of the `Paths` do not carry any credentials:
```python
https_bucket_path = Path("bfss://127.0.0.1:8888/service/bucket/some/file.tar.gz")
```

Still even if we run in such a scenario, as discussed back than I think the `Credential Store` abstraction can take care of this.

>                 Note: It is not clear yet if addition information will be required for the actual

Reviewer: ```suggestion
                Note: It is not clear yet if additional information will be required for the actual
```

                A Pathlike object for the given uri.
            """
            # type: LocalPath, BucketPath, Chroot ...
            # 
            # Note: based on the uri the factory should assemble the apropriate Pathlike object.
            # E.g.:
            type = _determine_type(path)
            facories = {
                "udf" = _create_udf_path,
                "bfs" = _create_bucket_path,
                "chroot" = _create_chroot_path,
            }
            factory = factory[type]
            return factory(args)

Examples
^^^^^^^^

.. code-block:: python

    from exasol.bucketfs import PathBuilder

    Path = PathBuilder(credentials)

    # Create different kinds of bucketfs paths
    udf_path = Path("bfsl://some/local/path/file.tar.gz")
    http_bucket_path = Path("bfs://127.0.0.1:8888/service/bucket/some/file.tar.gz")
    https_bucket_path = Path("bfss://127.0.0.1:8888/service/bucket/some/file.tar.gz")
    chroot_path = Path("bfss://127.0.0.1:8888/service/bucket/some/sub/subsub/file.tar.gz?chroot=/some/sub/")
    readonly_path = Path("bfsl://some/local/path/file.tar.gz?mode=ro")

Utilities
---------

.. code-block:: python

    Uri = str

>     def as_udf_path(path: Uri | BUcketPath) -> Pathlike:

Reviewer: ```suggestion
    def as_udf_path(path: Uri | BucketPath) -> Pathlike:
```

                Returns an iterator which can be used to read the contents of the path in chunks.

            Raises:
                NotAFileError: if the pathlike object does not point to a file.
                FileNotFoundError: If the file does not exist.
            """

        def write(data: ByteString | BinaryIO | Iterable[ByteString]):
            """
            Writes data to a this path. 

            After successfully writing to this path `exists` will yield true for this path.
            If the file already existed it will be overwritten.

            Args:
                data: which shall be writen to the path.

            Raises:
                NotAFileError: if the pathlike object is not a file path.
            """

        def rm():
            """
            Remove this file.

            Note:
                If `exists()` and is_file yields true for this path, the path will be deleted,
                otherwise exception will be thrown.

            Raises:
                FileNotFoundError: If the file does not exist.
            """

        def rmdir(recursive: bool = False):
            """
            Removes this directory.

>             Note: In order to stay close to the pathlib based `rmdir` the default won't delete
>                   non empty direcotries.

Reviewer 2: ```suggestion
            Note: In order to stay close to the pathlib, by default `rmdir` with `recursive`set to `False`
            won't delete non-empty diretories.
```

Reviewer 3: direCtories

Author: Thx @ both, in that case I will manually adjust ;)

            Args:

>                 recursive: if true the entrie directory and it's entire contents (files and subdirs)

Reviewer 2: ```suggestion
                recursive: if true the directory itself and it's entire contents (files and subdirs)
```

Reviewer 3: In this case it should be 'its'.

Author: Thx @ both, in that case I will manually adjust ;)

            Raises:
                FileNotFoundError: If the file does not exist.
                PermissionError: If recursive is false and the directory is not empty.
            """

        def joinpath(*pathsegements) -> Pathlike:
            """
            Calling this method is equivalent to combining the path with each of the given pathsegments in turn.

            Returns:
                A new pathlike object pointing the combined path.
            """

        def walk() -> Tuple[Pathlike, List[str], List[str]]:
            """
            Generate the file names in a directory tree by walking the tree either top-down or bottom-up.

            Note:
                Try to mimik https://docs.python.org/3/library/pathlib.html#pathlib.Path.walk as closely as possible,
                except the functionality associated with the parameters of the `pathlib` walk.

            Yields:
                A 3-tuple of (dirpath, dirnames, filenames).
            """

        def iterdir() -> Generator[Pathlike, None, None]:
            """
            When the path points to a directory, yield path objects of the directory contents.

            Note:

>                 For file like paths it will yield nothing.

Reviewer 2: ```suggestion
                If `path` points to a file then `iterdir()` will yield nothing.
```

            Yields:

>                 All direct childs of the pathlike object.

Reviewer 2: ```suggestion
                All direct children of the pathlike object.
```

            """
            Removes this directory.

            Note: In order to stay close to pathlib, by default `rmdir` with `recursive`

>                   set to `False` default won't delete non empty directories.

Reviewer 2: ```suggestion
                  set to `False` won't delete non-empty directories.
```
